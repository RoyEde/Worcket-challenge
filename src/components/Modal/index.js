import React, { useEffect, useRef, useState } from 'react';
import { bool, func } from 'prop-types';

import Button from '../Button';

const Modal = ({ loading, onAccept, onCancel, visible }) => {
  const ref = useRef();

  useEffect(() => {
    if (visible) ref.current && ref.current.focus();
    else setValue('');
  }, [visible]);

  const [value, setValue] = useState('');

  const handleAccept = () => {
    if (value) onAccept(value);
  };
  const handleChange = ({ target }) => setValue(target.value);
  const handleKeyPress = ({ key }) => {
    if (key === 'Enter') handleAccept();
  };

  return (
    visible && (
      <div className="modal" data-test="modal">
        <div className="modal__backdrop">
          <div className="modal__backdrop__body">
            <h2>Add an item</h2>
            <input
              className="modal__backdrop__body__input"
              data-test="add-input"
              onChange={handleChange}
              onKeyPress={handleKeyPress}
              ref={ref}
              type="text"
              value={value}
            />
            <div className="modal__backdrop__body__buttons">
              <Button
                aria-label="Close modal"
                data-test="cancel-item-button"
                disabled={loading}
                onClick={onCancel}
                >
                Close
              </Button>
              <Button
                aria-label={`Add item ${value}`}
                data-test="save-item-button"
                disabled={!value || loading}
                onClick={handleAccept}
                primary
              >
                Save
              </Button>
            </div>
          </div>
        </div>
      </div>
    )
  );
};

Modal.propTypes = {
  loading: bool,
  onAccept: func.isRequired,
  onCancel: func.isRequired,
  visible: bool
};

export default Modal;
