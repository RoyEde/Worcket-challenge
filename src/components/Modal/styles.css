.modal {
  height: 100vh;
  left: 0;
  position: fixed;
  top: 0;
  width: 100vw;
  z-index: 10;
}

.modal__backdrop {
  align-items: center;
  background-color: rgba(0, 0, 0, 0.4);
  display: flex;
  height: 100%;
  justify-content: center;
  padding: 1rem;
  width: 100%;
}

.modal__backdrop__body {
  background-color: #fff;
  border-radius: 5px;
  flex: 1;
  min-height: 12rem;
  max-width: 600px;
  padding: 0 1rem 0.5rem;
}

.modal__backdrop__body__input {
  appearance: none;
  border: 2px solid #dddddd;
  border-radius: 5px;
  font-family: inherit;
  font-size: 1rem;
  padding: 0.8rem 0.5rem;
  width: 100%;
}

.modal__backdrop__body__buttons {
  justify-content: space-between;
  display: flex;
  padding: 1rem 0;
}
