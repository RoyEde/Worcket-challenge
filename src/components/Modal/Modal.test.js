import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import Modal from './index';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Modal visible={false} onAccept={() => {}} onCancel={() => {}} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('snapshots match', () => {
  const component = renderer.create(<Modal visible onAccept={() => {}} onCancel={() => {}} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  
  tree.props.visible = false;
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  
  tree.props.visible = true;
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
