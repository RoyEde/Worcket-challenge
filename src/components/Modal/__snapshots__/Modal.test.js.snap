// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots match 1`] = `
<div
  className="modal"
>
  <div
    className="modal__backdrop"
  >
    <div
      className="modal__backdrop__body"
    >
      <h2>
        Add an item
      </h2>
      <input
        className="modal__backdrop__body__input"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value=""
      />
      <div
        className="modal__backdrop__body__buttons"
      >
        <button
          aria-label="Close modal"
          className="button   "
          onClick={[Function]}
        >
          Close
        </button>
        <button
          aria-label="Add item "
          className="button button--primary button--disabled "
          disabled={true}
          onClick={[Function]}
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshots match 2`] = `
<div
  className="modal"
>
  <div
    className="modal__backdrop"
  >
    <div
      className="modal__backdrop__body"
    >
      <h2>
        Add an item
      </h2>
      <input
        className="modal__backdrop__body__input"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value=""
      />
      <div
        className="modal__backdrop__body__buttons"
      >
        <button
          aria-label="Close modal"
          className="button   "
          onClick={[Function]}
        >
          Close
        </button>
        <button
          aria-label="Add item "
          className="button button--primary button--disabled "
          disabled={true}
          onClick={[Function]}
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshots match 3`] = `
<div
  className="modal"
>
  <div
    className="modal__backdrop"
  >
    <div
      className="modal__backdrop__body"
    >
      <h2>
        Add an item
      </h2>
      <input
        className="modal__backdrop__body__input"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value=""
      />
      <div
        className="modal__backdrop__body__buttons"
      >
        <button
          aria-label="Close modal"
          className="button   "
          onClick={[Function]}
        >
          Close
        </button>
        <button
          aria-label="Add item "
          className="button button--primary button--disabled "
          disabled={true}
          onClick={[Function]}
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;
